### PR TITLE
[DRAFT] Balancer Liquidity Update

### PR DESCRIPTION
**Pull request by viniabussafi**
Brief comments on the purpose of your changes:
- Joins the BPT prices table, allowing for a calculation of pool liquidity as pool_liquidity_usd
- Changes the amount_usd column to protocol_liquidity_usd, for clarity
- Adds Gnosis Liquidity spell, as well as the needed upstream tables
- Adds Avalanche Liquidity spell, as well as the needed upstream tables
- Adds a balancer V2 liquidity spell across all chains, in order to have easier TVL calculations

@mendesfabio @thetroyharris @markusbkoch 
**For Dune Engine V2**

I've checked that:

### General checks:
* [ ] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [ ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ ] if adding a new model, I added a test
* [ ] the filename is unique and ends with .sql
* [ ] each sql file is a select statement and has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`
* [ ] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [ ] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributor Dune usernames)

### Pricing checks:
* [ ] `coin_id` represents the ID of the coin on coinpaprika.com
* [ ] all the coins are active on coinpaprika.com (please remove inactive ones)

### Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

### Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

File: models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql [hunk 1]
```diff
@@ -165,41 +175,46 @@ zipped_balance_changes AS (
...
+        LEFT JOIN bpt_prices p3 ON p3.day = b.day AND p3.token = CAST(b.token as varchar(42))
         WHERE b.token != SUBSTRING(b.pool_id, 0, 42)
     ),
 
     pool_liquidity_estimates AS (
         SELECT
             b.day,
             b.pool_id,
-            SUM(b.amount_usd) / COALESCE(SUM(w.normalized_weight), 1) AS liquidity
+            SUM(b.protocol_liquidity_usd) / COALESCE(SUM(w.normalized_weight), 1) AS protocol_liquidity,
```
Comment (mendesfabio) [inline models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql:198]: shouldn't this be `pool_liquidity_usd`?

File: models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql [hunk 2]
```diff
@@ -165,41 +175,46 @@ zipped_balance_changes AS (
...
+            SUM(b.pool_liquidity_usd) / COALESCE(SUM(w.normalized_weight), 1) AS pool_liquidity
         FROM cumulative_usd_balance b
         LEFT JOIN {{ ref('balancer_v2_arbitrum_pools_tokens_weights') }} w ON b.pool_id = w.pool_id
         AND b.token = w.token_address
-        AND b.amount_usd > 0
+        AND b.protocol_liquidity_usd > 0
```
Comment (mendesfabio) [inline models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql:203]: same as previous comment

File: models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql [hunk 3]
```diff
@@ -165,41 +175,46 @@ zipped_balance_changes AS (
...
         AND c.day < b.day_of_next_change
-        LEFT JOIN {{ ref('tokens_erc20_legacy') }} t ON t.contract_address = b.token
+        LEFT JOIN {{ ref('tokens_erc20') }} t ON t.contract_address = b.token
         AND blockchain = "arbitrum"
         LEFT JOIN prices p1 ON p1.day = b.day
         AND p1.token = b.token
         LEFT JOIN dex_prices p2 ON p2.day <= c.day
         AND c.day < p2.day_of_next_change
         AND p2.token = b.token
+        LEFT JOIN bpt_prices p3 ON p3.day = b.day AND p3.token = CAST(b.token as varchar(42))
```
Comment (mendesfabio) [inline models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql:190]: not urgent, but we have to agree on the types of addresses to avoid having the need for casting everywhere

File: models/balancer/gnosis/balancer_gnosis_sources.yml [hunk 4]
```diff
@@ -3,7 +3,7 @@ version: 2
 sources:
   - name: balancer_v2_gnosis
     description: >
-      Decoded tables related to Balancer V2, an automated portfolio manager and trading platform, on Gnosis.
+      Decoded tables related to Balancer v2, an automated portfolio manager and trading platform, on gnosis.
```
Comment (mendesfabio) [inline models/balancer/gnosis/balancer_gnosis_sources.yml:6]: not a big deal, but I don't understand the chance to gnosis

File: models/balancer/gnosis/balancer_gnosis_sources.yml [hunk 5]
```diff
@@ -25,15 +25,15 @@ sources:
...
+            description: 'gnosis address for the liquidity pool used in transaction'
           - name: poolId
             description: 'Unique encoded identifier that refers to each pool'
           - name: specialization
             description: 'Pool specialization'
 
       - name: Vault_evt_Swap
         description: >
-          Decoded table related to the the contracts emitted by swaps in Balancer V2 pools.
+          Decoded table related to the contracts emitted by swaps in Balancer V2 pools.
```
Comment (mendesfabio) [inline models/balancer/gnosis/balancer_gnosis_sources.yml:36]: I assume this file was copied from other chains - could you fix the same typo on other sources yaml files?

File: models/balancer/gnosis/balancer_v2_gnosis_pools_tokens_weights.sql [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+--
+-- Balancer v2 Pools Tokens Weights
+--
+
+SELECT
+    registered.poolId AS pool_id,
+    tokens.token_address,
+    weights.normalized_weight / POWER(10, 18) AS normalized_weight
+FROM {{ source('balancer_v2_gnosis', 'Vault_evt_PoolRegistered') }} registered
+INNER JOIN {{ source('balancer_v2_gnosis', 'WeightedPoolV2Factory_call_create') }} call_create
```
Comment (mendesfabio) [inline models/balancer/gnosis/balancer_v2_gnosis_pools_tokens_weights.sql:25]: we're currently at WeightedV4 - have we sent new factories for decoding?

File: models/balancer/optimism/balancer_optimism_schema.yml [hunk 7]
```diff
@@ -265,18 +265,21 @@ models:
     columns:
       - name: day
         description: 'UTC event block time truncated to the day mark'
-      - *pool_id
+      - name: pool_id
```
Comment (mendesfabio) [inline models/balancer/optimism/balancer_optimism_schema.yml:270]: why did you remove the asterisk? from Dune docs:

> “&” is used for the first definition of a column and “*” thereafter will lead to a column with the same name in a different model inheriting the same name, description, and generic tests.

File: models/balancer/balancer_schema.yml [hunk 8]
```diff
@@ -172,4 +172,38 @@ models:
...
+            - token_symbol
+    columns:
+      - name: day
+        description: 'UTC event block time truncated to the day mark'
+      - name: pool_id
+      - name: pool_symbol
+        description: 'Symbol of the pool, consisting of the symbol of its tokens and their respective weights'
+      - name: blockchain 
+      - name: token_address
+      - name: token_symbol
```
Comment (mendesfabio) [inline models/balancer/balancer_schema.yml:200]: I think some of these columns have been declared before and you could inherit name, description, etc, using `*` mark

File: models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql [hunk 9]
```diff
@@ -52,6 +52,16 @@ WITH pool_labels AS (
             dex_prices_1
     ),
 
+    
+    bpt_prices AS(
+        SELECT 
+            date_trunc('day', HOUR) AS DAY,
+            contract_address AS token,
+            percentile(median_price, 0.5) AS bpt_price
+        FROM {{ ref('balancer_bpt_prices') }}
```
Comment (mendesfabio) [inline models/balancer/arbitrum/balancer_v2_arbitrum_liquidity.sql:61]: should filter by blockchain

**Review by mendesfabio — COMMENTED** [review]
I left some comments for you to fix @viniabussafi -- I tried to not be repetitive but note that some comments apply to all chains.

File: models/balancer/gnosis/balancer_v2_gnosis_pools_tokens_weights.sql (hunk 6 above)
Comment (viniabussafi) [reply models/balancer/gnosis/balancer_v2_gnosis_pools_tokens_weights.sql:25]: just sent.
